src/Plutarch/Extra/Value.hs

```diff
@@ -170,3 +199,568 @@ phasOnlyOneTokenOfCurrencySymbol = phoistAcyclic $
+    HRecToList rest SomeAssetClass =>
+    HRecToList
+        ('(name, AssetClass tag) ': rest)
+        SomeAssetClass
+    where
+    hrecToList (HCons (Labeled x) xs) = SomeAssetClass x : hrecToList xs
+
+-- | Associates given Symbol of AssetClass to PTagged tag PInteger
+type family OutputMatchValueAssets (ps :: [(Symbol, Type)]) (s :: S) where
+    OutputMatchValueAssets '[] _ = '[]
+    OutputMatchValueAssets ('(name, AssetClass tag) ': rest) s =
+        '( name
+         , Term
+            s
+            (PAsData (PTagged tag PInteger))
+         )
+            ': OutputMatchValueAssets rest s
+
+class MatchValueAssetReferences (input :: [(Symbol, Type)]) (s :: S) where
+    matchValueAssetReferences ::
+        Map SomeAssetClass (Term s (PAsData PInteger)) ->
+        HRec input ->
+        HRec (OutputMatchValueAssets input s)
+
+instance MatchValueAssetReferences '[] (s :: S) where
+    matchValueAssetReferences _ _ = HNil
+
+instance
+    forall
+        (name :: Symbol)
+        (classSymbol :: Symbol)
+        (is :: [(Symbol, Type)])
+        (s :: S).
+    MatchValueAssetReferences is s =>
+    MatchValueAssetReferences
+        ('(name, AssetClass classSymbol) ': is)
+        s
+    where
+    matchValueAssetReferences valueMap (HCons (Labeled cls) rest) =
+        HCons
+            ( Labeled @name
+                ( pdata $
+                    pcon $
+                        PTagged $
+                            pfromData (valueMap Map.! SomeAssetClass cls)
+                )
+            )
+            (matchValueAssetReferences valueMap rest)
+
+-- | Helper to construct the inner-mapping of a PValue
+mkSingleValue ::
+    forall (key :: KeyGuarantees) (tag :: Symbol) (s :: S).
+    Term
+        s
+        ( PAsData PCurrencySymbol :--> PAsData PTokenName :--> PTagged tag PInteger
+            :--> PBuiltinPair
+                    (PAsData PCurrencySymbol)
+                    (PAsData (PMap key PTokenName PInteger))
+        )
+mkSingleValue = phoistAcyclic $
+    plam $ \sym tk q ->
+        ppairDataBuiltin
+            # sym
+            # pdata (pcon $ PMap $ pfromList [ppairDataBuiltin # tk # pdata (pto q)])
+
+{- | Version of mkSingleValue with a haskell-level constant, hoisting with the
+ applied arguments
+-}
+mkSingleValue' ::
+    forall (tag :: Symbol) (k :: KeyGuarantees) (s :: S).
+    AssetClass tag ->
+    Term
+        s
+        ( PTagged tag PInteger
+            :--> PBuiltinPair
+                    (PAsData PCurrencySymbol)
+                    (PAsData (PMap k PTokenName PInteger))
+        )
+mkSingleValue' (AssetClass sym tk) =
+    phoistAcyclic $ mkSingleValue # pconstantData sym # pconstantData tk
+
+-- TODO: Remove me, once Seungheon's PR lands
+pfromList ::
+    forall (list :: (S -> Type) -> S -> Type) (a :: S -> Type) (s :: S).
+    (PIsListLike list a) =>
+    [Term s a] ->
+    Term s (list a)
+pfromList = foldr (\x xs -> pcons # x # xs) pnil
```

Comment:
Don't we have something like this already in `Plutarch.Extra.List`?

Reply:
Doesn't seem like it. @SeungheonOh, I know you mentioned you'd add this -- is there anything special you wanted to do here (optimizations, etc?) or is it fine to just yeet in `Plutarch.Extra.List` as is?

Reply:
I don't see any possible optimization here.